{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5276580f",
   "metadata": {},
   "source": [
    "**CLASS FASTAREADER**\n",
    "\n",
    "This class is for reading fasta files and creating a manipulatable file.\n",
    "\n",
    "Functions: \n",
    "    **\\__init__** - class constructor\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e273263b",
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Class: fastaReader\n",
    "#  \n",
    "# Purpose:class to read .fa files\n",
    "# \n",
    "#          \n",
    "# Author: Chris Nguyen\n",
    "# History:      cdn 10/10/21 Created\n",
    "#             \n",
    "########################################################################\n",
    "\n",
    "\n",
    "\n",
    "class fastaReader:\n",
    "    \n",
    "    \n",
    "    \n",
    "    def __init__(self, inFile):\n",
    "        '''\n",
    "        fastaReader constructor. Takes a file as input, creates a fastaReader object. \n",
    "        '''\n",
    "        self.file = inFile\n",
    "        self.fastaFile = []\n",
    "        self.headerFile = []\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        \n",
    "        with open(inFile) as fileR:\n",
    "            for line in fileR:\n",
    "                if line[0] == \">\":\n",
    "                    \n",
    "                    line = line.split()\n",
    "                    header = line[0][1 : len(line[0])]\n",
    "                    self.headerFile.append(header)\n",
    "                elif line[0] != \">\":\n",
    "                    \n",
    "                    sequence = line.rstrip(\"\\n\")\n",
    "                    self.fastaFile.append(sequence)\n",
    "                    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2380f98d",
   "metadata": {},
   "source": [
    "**CLASS RANDOMIZEDSEARCHMOTIFER**\n",
    "This class performs the actual searching. \n",
    "\n",
    "Functions:\n",
    "    \n",
    "    __init__(self, DNA, kmer, iterate, pseudo,  gibbs = False) : class constructor\n",
    "    makeNull(self) : makes the null model\n",
    "    countAndProfile(self, motifMatrix) : creates count and profile matrix\n",
    "    replaceMotif(self, profileMatrix, motifMatrix, index) : for gibbs sampling, replaces a random motif\n",
    "    fullCalc(self, profileMatrix) : replaces all motifs with better ones\n",
    "    scoreMatrix(self, profileMatrix) : scores a profile matrix\n",
    "    getConsensus(self, motifMatrix) : gets consensus sequence\n",
    "    performSearch(self) : does the full search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3448d2f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Class: randomizedSearchMotifer\n",
    "#  \n",
    "# Purpose:class to do the randomized searching for motifs\n",
    "# \n",
    "#          \n",
    "# Author: Chris Nguyen\n",
    "# History:      cdn 10/10/21 Created\n",
    "#             \n",
    "########################################################################\n",
    "class randomizedSearchMotifer:\n",
    "    \n",
    "    \n",
    "    def __init__(self, DNA, kmer, iterate, pseudo,  gibbs = False):\n",
    "        '''\n",
    "        Constructor for the randomizedSearchMotifer class.\n",
    "        Takes in parameters and stores them in member variables.\n",
    "        \n",
    "        Parameters:\n",
    "            DNA (list) - set of data to search from\n",
    "            kmer (int) - length of kmer to use\n",
    "            iterate (int) - how many times to do the algorithm\n",
    "            pseudo (float) - which value to use for pseudocount\n",
    "            gibbs (bool) - boolean to see if gibbs sampling should be done\n",
    "        '''\n",
    "        self.d = DNA \n",
    "        self.g = gibbs \n",
    "        self.k = kmer \n",
    "        self.i = iterate \n",
    "        self.p = pseudo \n",
    "        \n",
    "        \n",
    "        self.null = self.makeNull() \n",
    "        self.nullScore = 0 \n",
    "        \n",
    "        \n",
    "        \n",
    "    #creates null model to compare values   \n",
    "    def makeNull(self):\n",
    "        '''\n",
    "        Class function that makes the null model.\n",
    "        \n",
    "        Returns the null model as a list.\n",
    "        \n",
    "        '''\n",
    "        #initializes local variables\n",
    "        numA = numC = numT = numG = 0\n",
    "        #loop that counts how many instances of a base there are\n",
    "        for x in self.d:\n",
    "            numA += x.count('a')\n",
    "            numC += x.count('c')\n",
    "            numT += x.count('t')\n",
    "            numG += x.count('g')\n",
    "            \n",
    "        #after which, a total is yielded from the counts\n",
    "        total = numA+numC+numG+numT\n",
    "        pA = ((numA + self.p) / (total + self.p * len(self.d)))\n",
    "        pC = ((numC + self.p) / (total + self.p * len(self.d)))\n",
    "        pT = ((numT + self.p) / (total + self.p * len(self.d)))\n",
    "        pG = ((numG + self.p) / (total + self.p * len(self.d)))\n",
    "        \n",
    "        \n",
    "        return [pA, pC, pT, pG]\n",
    "        \n",
    "        \n",
    "        \n",
    "    def countAndProfile(self, motifMatrix):\n",
    "        '''\n",
    "        Class function that creates count matrix and subsequently creates profile matrix.\n",
    "        \n",
    "        Parameters: \n",
    "            motifMatrix (list) - motifs to count and profile\n",
    "            \n",
    "        Return:\n",
    "            profileMatrix (list) - profile matrix\n",
    "        '''\n",
    "        #initialization of local variables\n",
    "        profileMatrix = []\n",
    "        numA = 0\n",
    "        numC = 0\n",
    "        numT = 0\n",
    "        numG = 0\n",
    "        \n",
    "        #loop that counts instances of nucleotide\n",
    "        for j in range(0, self.k):\n",
    "            for i in motifMatrix:\n",
    "                if i[j] == 'a':\n",
    "                    numA += 1\n",
    "                elif i[j] == 'c':\n",
    "                    numC += 1\n",
    "                elif i[j] == 't':\n",
    "                    numT += 1\n",
    "                elif i[j] == 'g':\n",
    "                    numG += 1\n",
    "                    \n",
    "                    \n",
    "            #gets profile probability for a row, and then appends it\n",
    "            \n",
    "            numA = (numA + self.p) / ((4 * self.p) + len(motifMatrix))\n",
    "            numC = (numC + self.p) / ((4 * self.p) + len(motifMatrix))\n",
    "            numT = (numT + self.p) / ((4 * self.p) + len(motifMatrix))\n",
    "            numG = (numG + self.p) / ((4 * self.p) + len(motifMatrix))\n",
    "            profileMatrix.append([numA, numC, numT, numG])\n",
    "            \n",
    "            #zeros out the variables for re-use\n",
    "            numA = numC = numT = numG = 0\n",
    "        \n",
    "        return profileMatrix\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "   \n",
    "    def replaceMotif(self, profileMatrix, motifMatrix, index):\n",
    "        '''\n",
    "        Class function that replaces a random motif in the motif matrix with the most profile-supported one.\n",
    "        Only used for Gibbs Sampling.\n",
    "        \n",
    "        Parameters:\n",
    "            profileMatrix (list) - matrix of profile probabilities\n",
    "            motifMatrix (list) - matrix of motifs being used\n",
    "            index (int) - location of motif being replaced in motifMatrix\n",
    "            \n",
    "        Return:\n",
    "            motifMatrix (list) - motif matrix with one motif mutated\n",
    "        \n",
    "        '''\n",
    "        #initialization of local variables\n",
    "        bestKmer = ''\n",
    "        probNew = 0\n",
    "        \n",
    "        #loop chooses kmers for consideration\n",
    "        for i in range(0, (len(self.d[index]) - self.k + 1)):\n",
    "            kmer = self.d[index][i : (i+self.k) ]\n",
    "            prob = 1\n",
    "            \n",
    "            \n",
    "            #probability created by comparing it to profile matrix\n",
    "            for x in range(0, (len(kmer))):\n",
    "                if kmer[x].upper() == \"A\":\n",
    "                    \n",
    "                    prob = prob * profileMatrix[x][0]\n",
    "                elif kmer[x].upper() == \"C\":\n",
    "                    \n",
    "                    prob = prob * profileMatrix[x][1]\n",
    "                elif kmer[x].upper() == \"T\":\n",
    "                    \n",
    "                    prob = prob * profileMatrix[x][2]\n",
    "                elif kmer[x].upper() == \"G\":\n",
    "                    \n",
    "                    prob = prob * profileMatrix[x][3]\n",
    "            \n",
    "            #if it's better, it replaces the best one so far\n",
    "            if prob > probNew:\n",
    "                probNew = prob\n",
    "                \n",
    "                bestKmer = kmer\n",
    "            \n",
    "            \n",
    "        #replaces motif in matrix once overall best one is found\n",
    "        motifMatrix[index] = bestKmer\n",
    "        return motifMatrix\n",
    "        \n",
    "        \n",
    "        \n",
    "    #calculates k-mer probabilities for all k-mers in the matrix\n",
    "    #input takes profileMatrix and motifMatrix and returns a better motifMatrix\n",
    "    def fullCalc(self, profileMatrix):\n",
    "        \n",
    "        '''\n",
    "        Class function that replaces all motifs in the motif matrix with the most profile-supported one.\n",
    "        \n",
    "        \n",
    "        Parameters:\n",
    "            profileMatrix (list) - matrix of profile probabilities\n",
    "            \n",
    "            \n",
    "        Return:\n",
    "            motifMatrix (list) - new set of motifs\n",
    "        \n",
    "        '''\n",
    "        #initialization of local variables\n",
    "        bestKmer = ''\n",
    "        probBest = 0\n",
    "        motifMatrix = []\n",
    "       \n",
    "        #loop for choosing a string from the whole data set\n",
    "        for string in self.d:\n",
    "            #loop for picking all kmers from a string\n",
    "            for i in range(0, (len(string) - self.k + 1)):\n",
    "                \n",
    "                kmer = string[i : (i + self.k) ]\n",
    "                \n",
    "                \n",
    "                prob = 1\n",
    "                \n",
    "                \n",
    "                #sees what its probability is by comparing it to profile matrix\n",
    "                for x in range(0, (len(kmer) - 1 )):\n",
    "                    \n",
    "                    if kmer[x].upper() == \"A\":\n",
    "                        \n",
    "                        prob = prob * profileMatrix[x][0]\n",
    "                        \n",
    "                    elif kmer[x].upper() == \"C\":\n",
    "                        \n",
    "                        prob = prob * profileMatrix[x][1]\n",
    "                    elif kmer[x].upper() == \"T\":\n",
    "                        \n",
    "                        prob = prob * profileMatrix[x][2]\n",
    "                    elif kmer[x].upper() == \"G\":\n",
    "                        \n",
    "                        prob = prob * profileMatrix[x][3]\n",
    "                \n",
    "                #if better, replaces the best one so far\n",
    "                if prob > probBest:\n",
    "                    probBest = prob\n",
    "                    \n",
    "                    bestKmer = kmer\n",
    "                    \n",
    "            #best one gets put into the matrix and probability is zeroed out for next calculation\n",
    "            motifMatrix.append(bestKmer)\n",
    "            probBest = 0\n",
    "                    \n",
    "            \n",
    "            \n",
    "        #replaces motif in matrix \n",
    "        \n",
    "        return motifMatrix\n",
    "        \n",
    "        \n",
    "    def scoreMatrix(self, profileMatrix):\n",
    "        '''\n",
    "        Class function that scores a profile matrix using relative entropy.\n",
    "        \n",
    "        Parameters:\n",
    "            profileMatrix (list) - matrix of profile probabilities\n",
    "            \n",
    "        Return:\n",
    "            relEnt (float) - score of matrix\n",
    "        \n",
    "        '''\n",
    "        #initialization of local variable\n",
    "        relEnt = 0\n",
    "        #loop that calculates relative entropy by visiting all elements in matrix\n",
    "        for x in range(0, len(profileMatrix)):\n",
    "            for y in range(0, len(profileMatrix[x])):\n",
    "              \n",
    "                \n",
    "                relEnt += profileMatrix[x][y] * math.log2((profileMatrix[x][y]) / (self.null[y]))\n",
    "        \n",
    "        return relEnt\n",
    "        \n",
    "        \n",
    "        \n",
    "    def getConsensus(self, motifMatrix):\n",
    "        '''\n",
    "        Class function that produces a consensus sequence for a given motif matrix.\n",
    "        \n",
    "        Parameters:\n",
    "            motifMatrix (list) - matrix of motifs being used\n",
    "            \n",
    "        Return:\n",
    "            conString (str) - string containing consensus sequence\n",
    "        \n",
    "        '''\n",
    "        #initialization of local variables\n",
    "        numA = numC = numT = numG = 0\n",
    "        conString = ''\n",
    "        \n",
    "        #counts nucleotide occurence for motif matrix\n",
    "        for j in range(0, self.k):\n",
    "            for i in motifMatrix:\n",
    "                \n",
    "                if i[j] == 'a':\n",
    "                    numA += 1\n",
    "                elif i[j] == 'c':\n",
    "                    numC += 1\n",
    "                elif i[j] == 't':\n",
    "                    numT += 1\n",
    "                elif i[j] == 'g':\n",
    "                    numG += 1\n",
    "            \n",
    "            consCount = max(numA, numC, numT, numG)\n",
    "           \n",
    "            #if the largest expression number matches the count for a given base, it's written to the consensus\n",
    "            if consCount == numA:\n",
    "                \n",
    "                conString += 'A'\n",
    "            elif consCount == numC:\n",
    "                \n",
    "                conString += 'C'\n",
    "            elif consCount == numT:\n",
    "                \n",
    "                conString += 'T'\n",
    "            elif consCount == numG:\n",
    "                \n",
    "                conString += 'G'\n",
    "                    \n",
    "            #zeroes out variables for next round\n",
    "            numA = numC = numT = numG = 0\n",
    "\n",
    "        return conString\n",
    "                \n",
    "        \n",
    "        \n",
    "    def performSearch(self):\n",
    "        '''\n",
    "        Class function that does the entire search, using Gibbs if specified\n",
    "        \n",
    "            \n",
    "        Return:\n",
    "            finalScore (float) - final true best score for the best motifs chosen from the data\n",
    "            finalConsensus (str) - final true best consensus for the file\n",
    "            finalMotifs (list) - final true best set of motifs across all iterations\n",
    "        \n",
    "        '''\n",
    "\n",
    "        #initialization of local variables\n",
    "        finalScore = 0\n",
    "        finalConsensus = ''\n",
    "        finalMotifs = []\n",
    "        #for a given number of iterations, performs the search\n",
    "        for i in range(0, self.i):\n",
    "            #iteration-local variables\n",
    "            motifs = []\n",
    "            newMat = []\n",
    "        \n",
    "            #loops through each sequence in the DNA\n",
    "            for seq in self.d:\n",
    "\n",
    "                #makes indices for k-mers\n",
    "\n",
    "                start = random.randint(0, len(seq) - self.k)\n",
    "                end = start + self.k + 1 #this makes it inclusive\n",
    "                kmer = seq[start:end]\n",
    "\n",
    "                motifs.append(kmer)\n",
    "                \n",
    "\n",
    "            #then, puts them all into a set of motifs called \"bestMotifs\"\n",
    "            bestMotifs = motifs\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            \n",
    "            #now, we have everything we need to do either method.\n",
    "            #if self.g is True, do Gibbs\n",
    "            if self.g :\n",
    "\n",
    "                #for gibbs, the iterations are used again for the inner calculation \n",
    "                for i in range(0, self.i):\n",
    "\n",
    "                    \n",
    "                    #randomly choosing a motif to discard\n",
    "\n",
    "                    bString = random.randint(0, len(self.d) - 1) \n",
    "\n",
    "                    #make new matrix without the string to pass to function\n",
    "                    for h in range(0, len(self.d)):\n",
    "                        \n",
    "                        #if index is not equal to the random int, put into new matrix\n",
    "                        if h != bString:\n",
    "                            newMat.append( bestMotifs[h])\n",
    "                            \n",
    "                    #after, we pass into the function to make a profile\n",
    "                    profile = self.countAndProfile(newMat) \n",
    "\n",
    "                    #use the resulting profile to replace a motif\n",
    "                    motifs = self.replaceMotif(profile, motifs, bString)\n",
    "\n",
    "                    #if the score of the new matrix is better than the current best, it replaces the current best\n",
    "                    if self.scoreMatrix(self.countAndProfile(motifs)) > self.scoreMatrix(self.countAndProfile(bestMotifs)):\n",
    "                        bestMotifs = motifs\n",
    "                \n",
    "                bestProfile = self.countAndProfile(bestMotifs)\n",
    "\n",
    "                #at the end, the overall best ones get chosen and returned\n",
    "                finalConsensus = self.getConsensus(bestMotifs)\n",
    "                finalScore = self.scoreMatrix(bestProfile)\n",
    "                finalMotifs = bestMotifs\n",
    "                return finalScore, finalConsensus, finalMotifs\n",
    "\n",
    "\n",
    "            #if self.g is False, do the search without Gibbs\n",
    "            else:\n",
    "                \n",
    "                \n",
    "                bestProfile = self.countAndProfile(bestMotifs)\n",
    "\n",
    "\n",
    "                while True:\n",
    "\n",
    "                    #use profile to choose another set of motifs \n",
    "                    newMotifs = self.fullCalc(bestProfile)\n",
    "                    newProfile = self.countAndProfile(newMotifs)\n",
    "                    \n",
    "                    #if the new matrix score is better, it replaces the previous best\n",
    "                    if self.scoreMatrix(newProfile) > self.scoreMatrix(bestProfile):\n",
    "                        \n",
    "                        bestMotifs = newMotifs\n",
    "                        bestProfile = newProfile\n",
    "                      \n",
    "                    #otherwise, the best has been found\n",
    "                    else:\n",
    "                        \n",
    "                        \n",
    "                        bestMotifs = newMotifs\n",
    "                        bestProfile = newProfile\n",
    "                        bestConsensus = self.getConsensus(bestMotifs)\n",
    "                        bestScore = self.scoreMatrix(bestProfile)\n",
    "                        break\n",
    "\n",
    "                    \n",
    "                #if the best score from an iteration is better than the overall best, it gets replaced\n",
    "                if finalScore == 0 or finalScore < bestScore:\n",
    "                    \n",
    "                    finalMotifs = bestMotifs\n",
    "                    finalScore = bestScore\n",
    "                    finalConsensus = bestConsensus\n",
    "\n",
    "\n",
    "        return finalScore, finalConsensus , finalMotifs\n",
    "\n",
    "\n",
    "\n",
    "            \n",
    "            \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f07c3007",
   "metadata": {},
   "source": [
    "**CLASS COMMANDLINE**\n",
    "\n",
    "This class specifies a commandline to use for the program.\n",
    "\n",
    "Functions: \n",
    "\n",
    "    __init__(self, inOpts=None) : class constructor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ae99a910",
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Class: CommandLine\n",
    "#  \n",
    "# Purpose:class to represent a command line\n",
    "# \n",
    "#          \n",
    "# Author: Chris Nguyen\n",
    "# History:      cdn 10/10/21 Created\n",
    "#             \n",
    "########################################################################\n",
    "import argparse\n",
    "class CommandLine() :\n",
    "\n",
    "    \n",
    "    def __init__(self, inOpts=None) :\n",
    "        '''\n",
    "        Class constructor that creates the command line.\n",
    "        \n",
    "        '''\n",
    "\n",
    "\n",
    "     \n",
    "        \n",
    "        self.parser = argparse.ArgumentParser(description = 'Program prolog - a brief description of what this thing does', \n",
    "                                             epilog = 'Program epilog - some other stuff you feel compelled to say', \n",
    "                                             add_help = True, #default is True \n",
    "                                             prefix_chars = '-', \n",
    "                                             usage = '%(prog)s [options] -option1[default] <input >output' \n",
    "                                             )\n",
    "        #these arguments can be accessed or seen using \"args.(argument name)\" so args.list would hold the list of values passed to the option\n",
    "        self.parser.add_argument('-i', '--iterations', type=int, action = 'store', required = True, help='iterations to perform')\n",
    "        self.parser.add_argument('-k', '--motifLength', type=int, choices= range(5, 48), action = 'store', required = True,  help='length')\n",
    "        self.parser.add_argument('-p', '--pseudocount', type = float, action = 'store', required = True, help='designates pseudocount value')\n",
    "        self.parser.add_argument('-g', '--gibbs', action = 'store_true', help = 'performs gibbs sampling')\n",
    "        self.parser.add_argument('-m', '--printMotif', action = 'store_true', help = 'print motifs of consensus')\n",
    "      \n",
    "        if inOpts is None :\n",
    "            self.args = self.parser.parse_args()\n",
    "        else :\n",
    "            self.args = self.parser.parse_args(inOpts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ed2bd11",
   "metadata": {},
   "source": [
    "**MAIN**\n",
    "\n",
    "This contains main, which makes object instances and runs the search. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "42bf3254",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best score: 10.430719837701108\n",
      "Consensus: ACTTAAAAACCCA\n"
     ]
    }
   ],
   "source": [
    "########################################################################\n",
    "# File: Main\n",
    "#  \n",
    "# Purpose:main function for randomizedMotifSearch program\n",
    "# \n",
    "#          \n",
    "# Author: Chris Nguyen\n",
    "# History:      cdn 10/10/21 Created\n",
    "#             \n",
    "########################################################################\n",
    "import random\n",
    "import math\n",
    "def main(inFile, myCommandLine = None):\n",
    "    \n",
    "\n",
    "    import sys\n",
    "\n",
    "    cLine = CommandLine(myCommandLine)\n",
    "    kmer = cLine.args.motifLength\n",
    "    pseudo = cLine.args.pseudocount\n",
    "    iterate = cLine.args.iterations\n",
    "    gibbs = cLine.args.gibbs\n",
    "    printOut = cLine.args.printMotif\n",
    "    \n",
    "    \n",
    "    \n",
    "    reader = fastaReader(inFile)\n",
    "    DNA = reader.fastaFile\n",
    "    headers = reader.headerFile\n",
    "    \n",
    "    #need to read values in from command line to affect constructor\n",
    "    motifer = randomizedSearchMotifer(DNA, kmer, iterate, pseudo, gibbs)\n",
    "    score, consensus, motifs  = motifer.performSearch()\n",
    "    print(\"Best score: \" + str(score))\n",
    "    print(\"Consensus: \" + consensus)\n",
    "    \n",
    "    \n",
    "    if printOut:\n",
    "        for x in range(0, len(DNA)):\n",
    "            print(\"Name of organism: \" + headers[x])\n",
    "            print(\"DNA sequence: \" +  DNA[x])\n",
    "            print(\"Motif: \" + motifs[x]) \n",
    "            \n",
    "            \n",
    "if __name__ == \"__main__\":\n",
    "    main('allFilesCrisprs.fa', ['-i 1000', '-p 1', '-k 13'])\n",
    "          \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a10009d",
   "metadata": {},
   "source": [
    "**INSPECTIONS**\n",
    "\n",
    "*INSPECTED BY : DAN XU, CHRIS CONDON, IMMAAD MIR\n",
    "\n",
    "SUGGESTIONS:\n",
    "1. VARIABLE NAMES CAN BE A BIT MORE EXPLANATORY **(NOT TAKEN)**\n",
    "2. NAMES ARE A BIT LONG FOR SOME THINGS **(TAKEN)**\n",
    "3. TOO MUCH WHITESPACE MAKES IT HARD TO FOLLOW **(TAKEN)**\n",
    "4. Some of your if-else statements to update your ATCG counts could dangerous, since you end your last statement with else: --> update 'G'. This could be problematic in the case that there is an 'N' or non-canonical base in the sequence somewhere. **(TAKEN)**\n",
    "5. Naming is good other than fullCalc, change that one **(NOT TAKEN)**\n",
    "6. Re-order your classes for readability. I would start with the command line class, then the FastA class, then the dataMaker and RandomizedMotifSearch and finally end with the main. **(SOMEWHAT TAKEN)**\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
